### PR TITLE
qstat doesn't supply full machine name with -n when nodes added by IP address

### PR DESCRIPTION
**Pull request by npadole20**
#### Describe Bug or Feature
* When nodes are created using IP address, qstat -n does not show the full node name. Node name is truncated after the first dot.
* When we create nodes using FQDN, qstat -n does not display a valid exec_host string in the expected format of <host1>/<T1>*<P1>.

#### Describe Your Change

* Changed how "." is handled in a node name. If the node name is a valid IP address, the string is not truncated at the first ".".
* Concatenating the <T1>*<P1> information to the short name of the node, if nodes are created using FQDN.

#### Link to Design Doc

NA

#### Attach Test and Valgrind Logs/Output
[qstat_n.txt](https://github.com/PBSPro/pbspro/files/3904359/qstat_n.txt)
[qstat_n_win.txt](https://github.com/PBSPro/pbspro/files/3911762/qstat_n_win.txt)
[valgrind.log](https://github.com/PBSPro/pbspro/files/3926570/valgrind.log)

### CODE REVIEW

**Comment by npadole20** [comment]
[qstat_seg_fix.txt](https://github.com/PBSPro/pbspro/files/3911779/qstat_seg_fix.txt)

**Comment by npadole20** [comment]
PTL logs -
[qstat_fqdn_af.txt](https://github.com/PBSPro/pbspro/files/3916269/qstat_fqdn_af.txt)
[qstat_fqdn_bf.txt](https://github.com/PBSPro/pbspro/files/3916270/qstat_fqdn_bf.txt)
[qstat_ip_af.txt](https://github.com/PBSPro/pbspro/files/3916271/qstat_ip_af.txt)
[qstat_ip_bf.txt](https://github.com/PBSPro/pbspro/files/3916272/qstat_ip_bf.txt)

File: src/cmds/qstat.c [hunk 1]
```diff
@@ -598,49 +599,69 @@ trunc_value(char *value, int len, int wide_len, int wide)
...
+		strcpy(rest, nodes);
+	else
+		exit_qstat("out of memory");
+	while ((token = strtok_r(rest, "+", &rest))){
+		token_cp = (char*)malloc(strlen(token) + 1);
+		if(token_cp)
+			strcpy(token_cp, token);
+		else
+			exit_qstat("out of memory");
+		subtoken = strtok(token, "/");
```
Comment (mike0042) [inline src/cmds/qstat.c]: Please add a comment. Why are we searching for "/"? Is it meant to be a netmask?

File: src/cmds/qstat.c [hunk 2]
```diff
@@ -598,49 +599,69 @@ trunc_value(char *value, int len, int wide_len, int wide)
...
-					linebuf[i++] = *nodes++;
-			}
+	rest = malloc(strlen(nodes) + 1);
+	if(rest)
+		strcpy(rest, nodes);
+	else
+		exit_qstat("out of memory");
+	while ((token = strtok_r(rest, "+", &rest))){
+		token_cp = (char*)malloc(strlen(token) + 1);
+		if(token_cp)
```
Comment (mike0042) [inline src/cmds/qstat.c]: Add a space after if.

File: src/cmds/qstat.c [hunk 3]
```diff
@@ -598,49 +599,69 @@ trunc_value(char *value, int len, int wide_len, int wide)
...
+			strcpy(token_cp, token);
+		else
+			exit_qstat("out of memory");
+		subtoken = strtok(token, "/");
+		chunk = token_cp + strlen(subtoken);
+		ret = inet_pton(AF_INET, subtoken, &(check_ip.sin_addr));
+		if (ret == 1){
+			/* node name is an IP address */
+			node_name = subtoken;
+			strcat(node_name, chunk);
```
Comment (mike0042) [inline src/cmds/qstat.c]: Are you certain this string will be null terminated and that we'll never overrun the buffer?

File: src/cmds/qstat.c [hunk 4]
```diff
@@ -598,49 +599,69 @@ trunc_value(char *value, int len, int wide_len, int wide)
...
-				stp++;
-			nodes = stp++;
 		} else {
-			stp++;
+			/* Node name is not an IP address */
+			node_name = strtok(subtoken, ".");
+			strcat(node_name, chunk);
+		}
+		l = strlen(node_name);
+		if (i + l < 77){
```
Comment (mike0042) [inline src/cmds/qstat.c]: Add a space before the brace.

What is happening here? Why 77? Please add a comment.

File: src/cmds/qstat.c [hunk 5]
```diff
@@ -598,49 +599,69 @@ trunc_value(char *value, int len, int wide_len, int wide)
...
+			while (l) {
+				linebuf[i++] = *node_name++;
+				l--;
+			}
+			linebuf[i++] = '+';
```
Comment (mike0042) [inline src/cmds/qstat.c]: What is the significance of "+", and why are we not dealing with "/" as we did previously?

File: src/cmds/qstat.c [hunk 6]
```diff
@@ -598,49 +599,69 @@ trunc_value(char *value, int len, int wide_len, int wide)
...
+		} else {
+			/* flush line and start next */
+			linebuf[i] = '\0';
+			if (no_newl)
+				printf("%s", show_nonprint_chars(linebuf));
+			else
+				printf("   %s\n", show_nonprint_chars(linebuf));
```
Comment (mike0042) [inline src/cmds/qstat.c]: I think you can utilize a tertiary operator here.
```suggestion
				printf((no_newl ? "%s" : "   %s\n"), show_nonprint_chars(linebuf));
```

File: src/cmds/qstat.c [hunk 7]
```diff
@@ -598,49 +599,69 @@ trunc_value(char *value, int len, int wide_len, int wide)
...
+			while (l) {
+				linebuf[i++] = *node_name++;
+				l--;
+			}
+			linebuf[i++] = '+';
 		}
 	}
 	if (i != 0) {
-		linebuf[i] = '\0';
+		linebuf[--i] = '\0';
```
Comment (mike0042) [inline src/cmds/qstat.c]: This looks dangerous. The conditional should be (i > 0). Otherwise, you run the risk of data corruption.

File: test/tests/functional/pbs_qstat.py [hunk 8]
```diff
@@ -102,3 +102,42 @@ def test_qstat_qselect(self):
...
+        self.server.manager(MGR_CMD_DELETE, NODE, None, '')
+        ipaddr = socket.gethostbyname(self.mom.hostname)
+        attr_A = {'Mom': self.mom.hostname}
+        self.server.manager(MGR_CMD_CREATE, NODE, id=ipaddr, attrib=attr_A)
+        self.server.expect(NODE, {'state=free': 1})
+        j = Job(TEST_USER)
+        jid = self.server.submit(j)
+        self.server.expect(JOB, {'job_state': 'R'}, id=jid)
+        qstat_cmd = \
+            os.path.join(self.server.pbs_conf['PBS_EXEC'], 'bin', 'qstat')
```
Comment (mike0042) [inline test/tests/functional/pbs_qstat.py]: Please avoid the use of backslash for line continuation where possible in Python. Align the last two arguments one space to the right of the open parenthesis on the next line and pycodestyle will be happy.

File: test/tests/functional/pbs_qstat.py [hunk 9]
```diff
@@ -102,3 +102,42 @@ def test_qstat_qselect(self):
...
+
+    def test_qstat_n_fqdn(self):
+        self.server.manager(MGR_CMD_DELETE, NODE, None, '')
+        self.server.manager(MGR_CMD_CREATE, NODE, id=self.mom.hostname)
+        self.server.expect(NODE, {'state=free': 1})
+        j = Job(TEST_USER)
+        jid = self.server.submit(j)
+        self.server.expect(JOB, {'job_state': 'R'}, id=jid)
+        qstat_cmd = \
+            os.path.join(self.server.pbs_conf['PBS_EXEC'], 'bin', 'qstat')
```
Comment (mike0042) [inline test/tests/functional/pbs_qstat.py]: Please avoid the use of backslash for line continuation where possible in Python. Align the last two arguments one space to the right of the open parenthesis on the next line and pycodestyle will be happy.

File: test/tests/functional/pbs_qstat.py [hunk 10]
```diff
@@ -102,3 +102,42 @@ def test_qstat_qselect(self):
                               as_script=True)
         if ret['rc'] != 0:
             self.assertFalse(ret['err'][0], ret_msg)
+
+    def test_qstat_n_ip(self):
```
Comment (sujatapatnaik52) [inline test/tests/functional/pbs_qstat.py]: Can you add docstring for both the test cases?

File: src/cmds/qstat.c [hunk 11]
```diff
@@ -598,55 +599,83 @@ trunc_value(char *value, int len, int wide_len, int wide)
...
+		/* We are using '/' delimiter to extract the <host1> value
+		 * from <host1>/<T1>*<P1> string. We use the <host1> to identify if
+		 * the node is created using IP address.
+		 */
+		subtoken = strtok(token, "/");
+		chunk = token_cp + strlen(subtoken);
+		ret = inet_pton(AF_INET, subtoken, &(check_ip.sin_addr));
+		if (ret == 1) {
+			/* node name is an IP address */
+			pbs_asprintf(&node_name, "%s%s", subtoken, chunk);
```
Comment (sujatapatnaik52) [inline src/cmds/qstat.c]: you may need to free `node_name` after it's usage.

File: src/cmds/qstat.c [hunk 12]
```diff
@@ -598,55 +599,83 @@ trunc_value(char *value, int len, int wide_len, int wide)
...
+		} else {
+			/* Node name is not an IP address */
+			pbs_asprintf(&node_name, "%s%s", strtok(subtoken, "."), chunk);
+		}
+		len = strlen(node_name);
+		if (i + len < 77) {
```
Comment (sujatapatnaik52) [inline src/cmds/qstat.c]: As Mike suggested, please add the reason for 77.

File: src/cmds/qstat.c [hunk 13]
```diff
@@ -598,55 +599,83 @@ trunc_value(char *value, int len, int wide_len, int wide)
...
+			while (len) {
+				linebuf[i++] = *node_name++;
+				len--;
 			}
-
-			/* strip off domain name to keep string short */
-			while ((*stp != '+') && (*stp != ':') && (*stp != '\0'))
-				stp++;
-			nodes = stp++;
+			/* Appedning a  '+' here because we want to maintain the
```
Comment (sujatapatnaik52) [inline src/cmds/qstat.c]: Please correct the Typo for Appending.

File: src/cmds/qstat.c [hunk 14]
```diff
@@ -599,75 +599,83 @@ trunc_value(char *value, int len, int wide_len, int wide)
...
 		exit_qstat("out of memory");
-	while ((token = strtok_r(rest, "+", &rest))){
-		token_cp = (char*)malloc(strlen(token) + 1);
-		if(token_cp)
-			strcpy(token_cp, token);
-		else
+	/* The exec_host string has the format <host1>/<T1>*<P1>[+<host2>/<T2>*<P2>+... ].
+	 * We are using '+' delimiter to find each <host1>/<T1>*<P1> string.
+	 */
+	token = strtok_r(rest, "+", &saveptr);
```
Comment (mike0042) [inline src/cmds/qstat.c]: Thank you for adding the comment. Why are you using the reentrant version of strtok?

File: src/cmds/qstat.c [hunk 15]
```diff
@@ -599,75 +599,83 @@ trunc_value(char *value, int len, int wide_len, int wide)
...
 			/* flush line and start next */
 			linebuf[i] = '\0';
-			if (no_newl)
-				printf("%s", show_nonprint_chars(linebuf));
-			else
-				printf("   %s\n", show_nonprint_chars(linebuf));
+			printf((no_newl ? "%s" : "   %s\n"), show_nonprint_chars(linebuf));
 			i = 0;
-			while (l) {
+			while (len) {
```
Comment (mike0042) [inline src/cmds/qstat.c]: When a loop involves incrementing/decrementing a variable, it's almost always better suited as a for loop rather than a while loop.

File: src/cmds/qstat.c [hunk 16]
```diff
@@ -599,75 +599,83 @@ trunc_value(char *value, int len, int wide_len, int wide)
...
-		else
-			printf("   %s\n", show_nonprint_chars(linebuf));
+		printf((no_newl ? "%s\n" : "   %s\n"), show_nonprint_chars(linebuf));
 	} else if (no_newl)
 		printf("\n");
+	if (token_cp) {
+		free(token_cp);
+		token_cp = NULL;
+	}
+	if (rest) {
```
Comment (mike0042) [inline src/cmds/qstat.c]: No need to test if a variable is NULL before you free it.

File: src/cmds/qstat.c (hunk 14 above)
Comment (bhroam) [reply src/cmds/qstat.c]: I find using the reentrant version of strtok() best at all times.  It saves you from doing silly things like calling strtok() to break on fields, and then calling strtok() again to break into subfields.  There is no harm in using it.

File: src/cmds/qstat.c (hunk 12 above)
Comment (npadole20) [reply src/cmds/qstat.c]: @sujatapatnaik52 @mike0042 we were using 77 in this function originally. So I went with it.

**Comment by npadole20** [comment]
[qstat_n_review_2.txt](https://github.com/PBSPro/pbspro/files/3925946/qstat_n_review_2.txt)

File: src/cmds/qstat.c (hunk 12 above)
Comment (mike0042) [reply src/cmds/qstat.c]: I find it strange that the variable "i" was initialized to zero and hasn't changed to this point. Perhaps a remnant of some old code that never got properly cleaned up? I suggest we remove it here.

Please use the following:
```suggestion
		if (len < PBS_MAXNODENAME) {
```
Also, change the declaration of linebuf to be of size PBS_MAXNODENAME + 1.

File: src/cmds/qstat.c [hunk 17]
```diff
@@ -598,55 +599,81 @@ trunc_value(char *value, int len, int wide_len, int wide)
...
+			pbs_asprintf(&node_name, "%s%s", subtoken, chunk);
+		} else {
+			/* Node name is not an IP address */
+			pbs_asprintf(&node_name, "%s%s", strtok(subtoken, "."), chunk);
+		}
+		/* Backing up node_name as we are modifying the pointer further in the code */
+		node_name_bkp = node_name;
+		len = strlen(node_name);
+		if (i + len < 77) {
+			for (;len > 0; i++, len--) {
```
Comment (mike0042) [inline src/cmds/qstat.c]: Initialize i to zero in the for loop and remove the initialization earlier in this function. Better yet, use strncpy() to copy the string.

File: src/cmds/qstat.c (hunk 14 above)
Comment (mike0042) [reply src/cmds/qstat.c]: Ahh... thanks @bhroam. I overlooked the fact there is another call to strtok here.

File: src/cmds/qstat.c [hunk 18]
```diff
@@ -598,55 +599,81 @@ trunc_value(char *value, int len, int wide_len, int wide)
...
-				stp++;
-			nodes = stp++;
+			/* Appending a  '+' here because we want to maintain the
+			 * exec_host format i.e. <host1>/<T1>*<P1>[+<host2>/<T2>*<P2>+.
+			 */
+			linebuf[i++] = '+';
 		} else {
-			stp++;
+			/* flush line and start next */
+			linebuf[i] = '\0';
```
Comment (mike0042) [inline src/cmds/qstat.c]:
```suggestion
			linebuf[0] = '\0';
```

File: src/cmds/qstat.c [hunk 19]
```diff
@@ -598,55 +599,81 @@ trunc_value(char *value, int len, int wide_len, int wide)
...
+			printf((no_newl ? "%s" : "   %s\n"), show_nonprint_chars(linebuf));
+			for (i = 0;len > 0; i++, len--) {
```
Comment (mike0042) [inline src/cmds/qstat.c]: Again, you could use strncpy here.

File: src/cmds/qstat.c [hunk 20]
```diff
@@ -598,55 +599,81 @@ trunc_value(char *value, int len, int wide_len, int wide)
...
-			nodes = stp++;
+			/* Appending a  '+' here because we want to maintain the
+			 * exec_host format i.e. <host1>/<T1>*<P1>[+<host2>/<T2>*<P2>+.
+			 */
+			linebuf[i++] = '+';
 		} else {
-			stp++;
+			/* flush line and start next */
+			linebuf[i] = '\0';
+			printf((no_newl ? "%s" : "   %s\n"), show_nonprint_chars(linebuf));
```
Comment (mike0042) [inline src/cmds/qstat.c]: If linebuf is a zero length string, I don't see the point of this printf().

File: src/cmds/qstat.c [hunk 21]
```diff
@@ -598,55 +599,81 @@ trunc_value(char *value, int len, int wide_len, int wide)
...
+	struct sockaddr_in check_ip;
+	int ret = 0;
 
 	if ((nodes == NULL) || (*nodes == '\0'))
 		return;
 
 	i = 0;
-	stp = nodes;
-	while (*nodes != '\0') {
-		if (*stp == '[')
```
Comment (mike0042) [inline src/cmds/qstat.c]: It looks like you removed the check for brackets, which is what "i" was being used for. That's the source of my confusion. I think the result is a slight change in behavior, and I don't think that's what we want.

File: src/cmds/qstat.c (hunk 20 above)
Comment (mike0042) [reply src/cmds/qstat.c]: Ignore this. I think "i" needs to stay.

File: src/cmds/qstat.c (hunk 18 above)
Comment (mike0042) [reply src/cmds/qstat.c]: Ignore this. I think "i" needs to stay.

File: src/cmds/qstat.c (hunk 17 above)
Comment (mike0042) [reply src/cmds/qstat.c]: Ignore this.

File: src/cmds/qstat.c (hunk 12 above)
Comment (mike0042) [reply src/cmds/qstat.c]: We may still need "i" here, but do please make the PBS_MAXNODENAME change.

Comment (npadole20) [reply src/cmds/qstat.c]: @hirenvadalia explained the value 77 comes from CHAR_LINE_LIMIT which has the value 78. It is not related to PBS_MAXNODENAME. I will change 77 and linebuf[78] to use CHAR_LINE_LIMIT - 1 and CHAR_LINE_LIMIT  respectively.

File: src/cmds/qstat.c (hunk 21 above)
Comment (npadole20) [reply src/cmds/qstat.c]: The check here for '[' and ']' made sure that we did not remove the brackets from the the <host> of the exec_host string (<host>/<t1>*<p1> + <host1>/<t2>*<p2>). After the fix too, we are retaining the brackets. We are not specifically checking for brackets because if <host> is not a valid IP address, we are using the value of <host> as is.
I have attached the log which shows the behavior of qstat -n before and after the fix. 
[qstat_with_sq_bracket.txt](https://github.com/PBSPro/pbspro/files/3931568/qstat_with_sq_bracket.txt)

File: src/cmds/qstat.c (hunk 19 above)
Comment (npadole20) [reply src/cmds/qstat.c]: I will have to change all place to use str* function. I would prefer to keep it as it is if that's okay.

File: src/cmds/qstat.c [hunk 22]
```diff
@@ -598,55 +599,81 @@ trunc_value(char *value, int len, int wide_len, int wide)
...
+			/* node name is an IP address */
+			pbs_asprintf(&node_name, "%s%s", subtoken, chunk);
+		} else {
+			/* Node name is not an IP address */
+			pbs_asprintf(&node_name, "%s%s", strtok(subtoken, "."), chunk);
+		}
+		/* Backing up node_name as we are modifying the pointer further in the code */
+		node_name_bkp = node_name;
+		len = strlen(node_name);
+		if (i + len < CHAR_LINE_LIMIT - 1) {
```
Comment (mike0042) [inline src/cmds/qstat.c]: The variable "i" was initialized to zero at the start of the function and has remains unchanged. It's not needed here. Also, please add parenthesis around (CHAR_LINE_LIMIT - 1).

File: src/cmds/qstat.c [hunk 23]
```diff
@@ -598,55 +599,81 @@ trunc_value(char *value, int len, int wide_len, int wide)
...
+			for (;len > 0; i++, len--) {
```
Comment (mike0042) [inline src/cmds/qstat.c]: This is the first use of i in this function. Please initialize it to zero and remove the prior initialization.

File: src/cmds/qstat.c (hunk 19 above)
Comment (mike0042) [reply src/cmds/qstat.c]: That's fine, but please add a space after the first semicolon.

File: src/cmds/qstat.c (hunk 22 above)
Comment (npadole20) [reply src/cmds/qstat.c]: Every "node_name" we read we append it to "linebuf" till we don't reach the limit of (CHAR_LINE_LIMIT - 1). So we need "i" here to keep track of what was the index when we updated "linebuf" with the "node_name" value from the previous iteration.

File: src/cmds/qstat.c (hunk 23 above)
Comment (npadole20) [reply src/cmds/qstat.c]: If I initialize "i" to zero here, we will overwrite the value of "linebuf" from the previous iteration.

File: src/cmds/qstat.c (hunk 22 above)
Comment (mike0042) [reply src/cmds/qstat.c]: OK, I get it now. Thanks for the explanation.

**Comment by npadole20** [comment]
@mike0042 thank you.
